### PR TITLE
Support opt save load

### PR DESCRIPTION
**Pull request by pangengzheng**
### PR types

Others
### PR changes

Others
### Description

card-78318
support semi-auto optimizer save load

### CODE REVIEW

File: python/paddle/distributed/checkpoint/load_state_dict.py [hunk 1]
```diff
@@ -399,6 +405,8 @@ def load_state_dict(
         paddle.distributed.init_parallel_env()
 
     rank_to_files = get_rank_to_files(path, state_dict, process_group, use_dist)
+    if len(rank_to_files) <= 0:
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/load_state_dict.py:408]: <0 is impossible?

File: python/paddle/distributed/checkpoint/load_state_dict.py [hunk 2]
```diff
@@ -55,6 +55,12 @@ def get_rank_to_files(path, state_dict, process_group, use_dist):
             if local_tensor_index.tensor_key in state_dict:
                 necessary_files.append(file_name)
     necessary_data_files_set = set(necessary_files)
+    if len(necessary_data_files_set) <= 0:
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/load_state_dict.py:58]: <0 is impossible?

File: python/paddle/distributed/auto_parallel/api.py [hunk 3]
```diff
@@ -495,6 +495,83 @@ def step(self):
...
+            param_list = self._inner_opt._parameter_list
+        for param in param_list:
+            if param.stop_gradient:
+                continue
+            if hasattr(param, "main_grad"):
+                if param.main_grad is not None:
+                    return state_dict
+            else:
+                if param.grad is not None:
+                    return state_dict
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py:519]: I think it is a little bit trick

File: python/paddle/distributed/checkpoint/load_state_dict.py (hunk 2 above)
Comment (pangengzheng) [reply python/paddle/distributed/checkpoint/load_state_dict.py:58]: 嗯是不会有，但是也不影响正确性，习惯性这样写了。可修改

File: python/paddle/distributed/auto_parallel/api.py (hunk 3 above)
Comment (pangengzheng) [reply python/paddle/distributed/auto_parallel/api.py:519]: 暂时没有想到更好的做法，这里就先考虑这样，后面再思考是否有更好做法

**Review by zhiqiu — APPROVED** [review]
LGTM, to be refined in next pr

**Pull request merged**